File: pkg/pwrgrep/testdata/fixtures/java/SideEffects.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.net.Socket;
import java.sql.Statement;

public class SideEffects {
    void f(Statement stmt) throws Exception {
        // ruleid: java-side-effects
        Runtime.getRuntime().exec("ls");
        // ruleid: java-side-effects
        Files.delete(Path.of("tmp"));
        // ruleid: java-side-effects
        new Socket("example.com", 80);
        // ruleid: java-side-effects
        stmt.executeQuery("SELECT 1");
        // ok: java-side-effects
        System.getenv("HOME");
    }
}

File: pkg/pwrgrep/testdata/fixtures/python/side-effects.py
import os
import shutil
import subprocess

import requests


def f(cursor):
    # ruleid: python-side-effects
    os.system("ls")
    # ruleid: python-side-effects
    subprocess.run(["ls", "-l"])
    # ruleid: python-side-effects
    shutil.rmtree("tmp")
    # ruleid: python-side-effects
    with open("out.txt", "w") as out:
        pass
    # ok: python-side-effects
    with open("in.txt") as f:
        pass
    # ruleid: python-side-effects
    requests.get("https://example.com")
    # ruleid: python-side-effects
    cursor.execute("SELECT 1")
    # ok: python-side-effects
    os.getenv("HOME")
